apps/api/src/services/portfolio-snapshotter.service.ts

```diff
@@ -62,26 +64,151 @@ export class PortfolioSnapshotter {
+      );
+      const newPrices =
+        await this.priceTracker.getBulkPrices(tokensNeedingPrices);
+
+      // Merge new prices into our map (preserving existing successful prices)
+      for (const [tokenAddress, priceReport] of newPrices) {
+        // Only update if we got a successful price (not null)
+        if (priceReport !== null) {
+          priceMap.set(tokenAddress, priceReport);
+        } else if (!priceMap.has(tokenAddress)) {
+          // Set to null if we don't have it yet (to track that we tried)
+          priceMap.set(tokenAddress, null);
+        }
+      }
+
+      // Step 3: Identify tokens that still need prices after batch
+      const failedTokens: string[] = [];
+      for (const balance of balances) {
+        const priceReport = priceMap.get(balance.tokenAddress);
+        // Only count as failed if the fetch actually failed (null), not if price is 0
+        if (priceReport === null && balance.amount > 0) {
+          failedTokens.push(balance.tokenAddress);
+        }
+      }
+
+      // If no failures, we're done!
+      if (failedTokens.length === 0) {
+        repositoryLogger.debug(
+          `[PortfolioSnapshotter] All prices fetched successfully for agent ${agentId} on attempt ${attemptNumber}`,
+        );
+        break;
+      }
+
+      // Log remaining failures for this attempt
+      if (attemptNumber === maxRetries + 1) {
+        repositoryLogger.error(
+          `[PortfolioSnapshotter] Failed to fetch prices for ${failedTokens.length} tokens after ${maxRetries + 1} attempts for agent ${agentId}`,
+        );
+      } else {
+        repositoryLogger.debug(
+          `[PortfolioSnapshotter] ${failedTokens.length} tokens still missing prices for agent ${agentId}, will retry entire batch`,
+        );
+      }
+    }
+
+    // Step 3: Calculate total value
     let totalValue = 0;
+    let pricesFetched = 0;
+    let pricesFailed = 0;
+    const missingPriceTokens: string[] = [];
 
     for (const balance of balances) {
-      const priceResult = await this.priceTracker.getPrice(
-        balance.tokenAddress,
-        undefined, // Let PriceTracker determine blockchain type
-        balance.specificChain as SpecificChain,
-      );
+      // Skip zero balances
+      if (balance.amount === 0) {
+        continue;
+      }
 
-      if (priceResult) {
+      const priceResult = priceMap.get(balance.tokenAddress);
+      if (priceResult !== null && priceResult !== undefined) {
+        // We have a valid PriceReport (fetch succeeded)
         const valueUsd = balance.amount * priceResult.price;
         totalValue += valueUsd;
+        pricesFetched++;
       } else {
-        repositoryLogger.warn(
-          `[PortfolioSnapshotter] No price available for token ${balance.tokenAddress}, excluding from portfolio snapshot`,
-        );
+        // Price fetch failed (null) or token wasn't in the response (undefined)
```

Comment:
We're basically rechecking if there are any prices missing, which we already did in the retry loop above.  I think if we do the change I just suggested in another comment, we can fail fast as soon as we come out of the retry loop if there are still tokens missing, and can then assume we have prices for everything for the rest of this function